### PR TITLE
Enable Flash Attention with bool padding mask and unit tests

### PR DESCRIPTION
**Pull request by titaiwangms**
This pull request refactors the attention masking logic to use a new, more efficient padding mask for the ONNX Attention operator, leveraging its built-in causal masking. The main changes introduce the `create_padding_mask` utility, update model and attention code to use it, and add comprehensive tests to ensure correctness and efficiency.

**Attention Masking Refactor:**

* Added `create_padding_mask` in `src/mobius/components/_common.py`: This function generates a boolean padding mask for the ONNX Attention op, allowing the use of `is_causal=1` for native causal masking and enabling Flash Attention eligibility.
* Updated `_apply_attention` in `src/mobius/components/_attention.py` to always set `is_causal=1`, so the attention mask now only encodes padding (not causality), simplifying mask construction and improving performance. [[1]](diffhunk://#diff-e933c016bf0b73c8a42773e9b572aaee56dfebd1ad4e810978098e60fd8f338fL60-R74) [[2]](diffhunk://#diff-e933c016bf0b73c8a42773e9b572aaee56dfebd1ad4e810978098e60fd8f338fL125-R136) [[3]](diffhunk://#diff-e933c016bf0b73c8a42773e9b572aaee56dfebd1ad4e810978098e60fd8f338fR147)
* Replaced `create_attention_bias` with `create_padding_mask` in model code (`src/mobius/models/base.py`), updating logic and comments to reflect the new masking approach. [[1]](diffhunk://#diff-0bbc1868b3fa1da990c2298d6cb7dff2785d024d765ef633cd574b5afd408db6L33-R33) [[2]](diffhunk://#diff-0bbc1868b3fa1da990c2298d6cb7dff2785d024d765ef633cd574b5afd408db6L87-R96)

**API and Import Updates:**

* Added `create_padding_mask` to public imports in `src/mobius/components/__init__.py` and test imports. [[1]](diffhunk://#diff-8eacab4d5682bac795dd69c12af6cfb6e6ac9198de7bf37ed06fc748a4682d0cR95) [[2]](diffhunk://#diff-8eacab4d5682bac795dd69c12af6cfb6e6ac9198de7bf37ed06fc748a4682d0cR132) [[3]](diffhunk://#diff-a7e5f8e5cb3674b380e1a57ddfb04abbcd361ba1c2455fc6f9f9bcc553354cd8R15)

**Testing Improvements:**

* Added `TestCreatePaddingMask` test class to `src/mobius/components/_common_test.py`, verifying correct mask creation for both 2D and 3D inputs, and ensuring the new mask uses fewer ops than the old bias method.

These changes simplify attention masking, improve efficiency (especially for Flash Attention), and ensure correctness through targeted tests.

### CODE REVIEW

**Comment by titaiwangms** [comment]
(1) Does rewrite rules still work (optimization)?
(2) Is there any model still using create attention mask (additive bias)?

**Review by justinchuby — APPROVED** [review]

**Comment by titaiwangms** [comment]
> (1) Does rewrite rules still work (optimization)? (2) Is there any model still using create attention mask (additive bias)?

<html>
<body>
<strong style="box-sizing: border-box; border: 0px solid; margin: 0px; padding: 0px; font-weight: bolder; color: rgb(209, 213, 219); font-family: &quot;JetBrains Mono&quot;, ui-monospace, monospace; font-size: 14px; font-style: normal; font-variant-ligatures: common-ligatures contextual; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: pre-wrap; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Q1: GQA rewrite still works perfectly.</strong><span style="box-sizing: border-box; border: 0px solid; margin: 0px; padding: 0px; color: rgb(209, 213, 219); font-family: &quot;JetBrains Mono&quot;, ui-monospace, monospace; font-size: 14px; font-style: normal; font-variant-ligatures: common-ligatures contextual; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: pre-wrap; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"> The rewrite rule is dtype-agnostic — </span><code class="bg-th-bg-muted px-1 rounded text-yellow-600 dark:text-yellow-300" style="box-sizing: border-box; border: 0px solid; margin: 0px; padding: 0px; font-family: &quot;JetBrains Mono&quot;, ui-monospace, monospace; font-feature-settings: normal; font-variation-settings: normal; font-size: 14px; border-radius: 0.25rem; background-color: rgb(28, 33, 40); padding-inline: 4px; color: oklch(0.905 0.182 98.111); font-style: normal; font-variant-ligatures: common-ligatures contextual; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: pre-wrap; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">attention_bias</code><span style="box-sizing: border-box; border: 0px solid; margin: 0px; padding: 0px; color: rgb(209, 213, 219); font-family: &quot;JetBrains Mono&quot;, ui-monospace, monospace; font-size: 14px; font-style: normal; font-variant-ligatures: common-ligatures contextual; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: pre-wrap; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"> is a positional wildcard with no dtype check. More importantly, the rewrite </span><strong style="box-sizing: border-box; border: 0px solid; margin: 0px; padding: 0px; font-weight: bolder; color: rgb(209, 213, 219); font-family: &quot;JetBrains Mono&quot;, ui-monospace, monospace; font-size: 14px; font-style: normal; font-variant-ligatures: common-ligatures contextual; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: pre-wrap; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">discards the mask entirely</strong><span style="box-sizing: border-box; border: 0px solid; margin: 0px; padding: 0px; color: rgb(209, 213, 219); font-family: &quot;JetBrains Mono&quot;, ui-monospace, monospace; font-size: 14px; font-style: normal; font-variant-ligatures: common-ligatures contextual; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: pre-wrap; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"> and reconstructs padding info from </span><code class="bg-th-bg-muted px-1 rounded text-yellow-600 dark:text-yellow-300" style="box-sizing: border-box; border: 0px solid; margin: 0px; padding: 0px; font-family: &quot;JetBrains Mono&quot;, ui-monospace, monospace; font-feature-settings: normal; font-variation-settings: normal; font-size: 14px; border-radius: 0.25rem; background-color: rgb(28, 33, 40); padding-inline: 4px; color: oklch(0.905 0.182 98.111); font-style: normal; font-variant-ligatures: common-ligatures contextual; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: pre-wrap; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">attention_mask</code><span style="box-sizing: border-box; border: 0px solid; margin: 0px; padding: 0px; color: rgb(209, 213, 219); font-family: &quot;JetBrains Mono&quot;, ui-monospace, monospace; font-size: 14px; font-style: normal; font-variant-ligatures: common-ligatures contextual; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: pre-wrap; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"> graph input as </span><code class="bg-th-bg-muted px-1 rounded text-yellow-600 dark:text-yellow-300" style="box-sizing: border-box; border: 0px solid; margin: 0px; padding: 0px; font-family: &quot;JetBrains Mono&quot;, ui-monospace, monospace; font-feature-settings: normal; font-variation-settings: normal; font-size: 14px; border-radius: 0.25rem; background-color: rgb(28, 33, 40); padding-inline: 4px; color: oklch(0.905 0.182 98.111); font-style: normal; font-variant-ligatures: common-ligatures contextual; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: pre-wrap; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">seqlens_k</code><span style="box-sizing: border-box; border: 0px solid; margin: 0px; padding: 0px; color: rgb(209, 213, 219); font-family: &quot;JetBrains Mono&quot;, ui-monospace, monospace; font-size: 14px; font-style: normal; font-variant-ligatures: common-ligatures contextual; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: pre-wrap; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">/</span><code class="bg-th-bg-muted px-1 rounded text-yellow-600 dark:text-yellow-300" style="box-sizing: border-box; border: 0px solid; margin: 0px; padding: 0px; font-family: &quot;JetBrains Mono&quot;, ui-monospace, monospace; font-feature-settings: normal; font-variation-settings: normal; font-size: 14px; border-radius: 0.25rem; background-color: rgb(28, 33, 40); padding-inline: 4px; color: oklch(0.905 0.182 98.111); font-style: normal; font-variant-ligatures: common-ligatures contextual; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: pre-wrap; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">total_seq_len</code><span style="box-sizing: border-box; border: 0px solid; margin: 0px; padding: 0px; color: rgb(209, 213, 219); font-family: &quot;JetBrains Mono&quot;, ui-monospace, monospace; font-size: 14px; font-style: normal; font-variant-ligatures: common-ligatures contextual; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: pre-wrap; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">. Both old float and new bool mask subgraphs become dead code after rewrite.

</span><strong style="box-sizing: border-box; border: 0px solid; margin: 0px; padding: 0px; font-weight: bolder; color: rgb(209, 213, 219); font-family: &quot;JetBrains Mono&quot;, ui-monospace, monospace; font-size: 14px; font-style: normal; font-variant-ligatures: common-ligatures contextual; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: pre-wrap; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Q2: Why 20+ models don't use padding_mask:</strong><span style="box-sizing: border-box; border: 0px solid; margin: 0px; padding: 0px; color: rgb(209, 213, 219); font-family: &quot;JetBrains Mono&quot;, ui-monospace, monospace; font-size: 14px; font-style: normal; font-variant-ligatures: common-ligatures contextual; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: pre-wrap; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">
</span><div class="my-2 overflow-x-auto" style="box-sizing: border-box; border: 0px solid; margin: 0px; padding: 0px; margin-block: 8px; overflow-x: auto; color: rgb(209, 213, 219); font-family: &quot;JetBrains Mono&quot;, ui-monospace, monospace; font-size: 14px; font-style: normal; font-variant-ligatures: common-ligatures contextual; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: pre-wrap; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">
Category | Models | Reason
-- | -- | --
Could switch (16) | OLMo, Granite, InternLM, GPT2, DeepSeek, MoE variants, Bamba, Jamba, Qwen3.5/Next/VL/TTS/ASR, Mllama, Falcon RoPE, Phi4-MM | Own TextModel class for structural reasons (norms, MoE, SSM) but masking needs are identical — one-line change each
Cannot switch (6) | Gemma2, Gemma3, Gemma3n, SmolLM3 | Need sliding_window parameter baked into float mask
  | Gemma v1 | Softcap + own Attention class without is_causal
  | Falcon ALiBi | ALiBi is positional encoding, fundamentally different from a mask

</div><span style="box-sizing: border-box; border: 0px solid; margin: 0px; padding: 0px; color: rgb(209, 213, 219); font-family: &quot;JetBrains Mono&quot;, ui-monospace, monospace; font-size: 14px; font-style: normal; font-variant-ligatures: common-ligatures contextual; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: pre-wrap; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">

</span><strong style="box-sizing: border-box; border: 0px solid; margin: 0px; padding: 0px; font-weight: bolder; color: rgb(209, 213, 219); font-family: &quot;JetBrains Mono&quot;, ui-monospace, monospace; font-size: 14px; font-style: normal; font-variant-ligatures: common-ligatures contextual; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: pre-wrap; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Bottom line:</strong><span style="box-sizing: border-box; border: 0px solid; margin: 0px; padding: 0px; color: rgb(209, 213, 219); font-family: &quot;JetBrains Mono&quot;, ui-monospace, monospace; font-size: 14px; font-style: normal; font-variant-ligatures: common-ligatures contextual; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: pre-wrap; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"> PR correctly scoped to base.TextModel as a safe starting point. 16 models could migrate in follow-up PRs (mechanical change). 6 models genuinely need the float bias.

Standing by for further questions or next tasks.</span>
</body>
</html>
